File: 4. Binary Tree/5_tree_min_value.java
import java.util.Queue;
import java.util.LinkedList;

class Node<T> {
    T val;
    Node<T> left;
    Node<T> right;

    public Node(T val) {
        this.val = val;
        this.left = null;
        this.right = null;
    }
}

class Source {
    public static Double treeMinValue(Node<Double> root) {

        double min = Double.MAX_VALUE;
        Queue<Node<Double>> queue = new LinkedList<>();
        Node<Double> current = root;

        if (root != null) {
            queue.add(current);
        }

        while (!queue.isEmpty()) {
            current = queue.remove();

            if (current.val < min) {
                min = current.val;
            }

            if (current.left != null) {
                queue.add(current.left);
            }

            if (current.right != null) {
                queue.add(current.right);
            }
        }

        return min;
    }
}

File: 4. Binary Tree/5_tree_min_value.py
from collections import deque

class Node:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None

def tree_min_value(root):
    if root is None:
        return float('inf')

    min_val = float('inf')
    queue = deque([root])
    while queue:
        current = queue.popleft()
        if current.val < min_val:
            min_val = current.val
        if current.left:
            queue.append(current.left)
        if current.right:
            queue.append(current.right)

    return min_val
